blob/azureblob/azureblob.go

```diff
@@ -816,11 +822,43 @@ func (b *bucket) ListPaged(ctx context.Context, opts *driver.ListOptions) (*driv
 	return page, nil
 }
 
+func (b *bucket) getDelegationCredentials(ctx context.Context) (*azblob.UserDelegationCredential, error) {
+	validity := 48 * time.Hour
+	currentTime := time.Now().UTC()
+	expires := currentTime.Add(validity)
+	b.opts.CredentialExpiration = expires
+	keyInfo := azblob.NewKeyInfo(currentTime, expires)
+	delegationCredentials, err := b.serviceURL.GetUserDelegationCredential(ctx, keyInfo, nil, nil)
+
+	if err != nil {
+		return nil, fmt.Errorf("azureblob.OpenBucket: error while retrieving user delegation credential: %s", err)
+	}
+
+	return &delegationCredentials, nil
+}
+
 // SignedURL implements driver.SignedURL.
 func (b *bucket) SignedURL(ctx context.Context, key string, opts *driver.SignedURLOptions) (string, error) {
+
 	if b.opts.Credential == nil {
-		return "", gcerr.New(gcerr.Unimplemented, nil, 1, "azureblob: to use SignedURL, you must call OpenBucket with a non-nil Options.Credential")
+		var err error
+		isMSIEnvironment := adal.MSIAvailable(ctx, adal.CreateSender())
+		if isMSIEnvironment {
+			if b.opts.Credential, err = b.getDelegationCredentials(ctx); err != nil {
```

Comment:
Let's store the delegation credentials (and expiration) on `bucket` instead of in `Options`.

Reply:
Ping? `Options` should be immutable.

Reply:
oh lord i missed the first part of this!

sorry!